compiler/rustc_metadata/src/creader.rs

```diff
@@ -1011,7 +1011,12 @@ impl<'a> CrateLoader<'a> {
                     CrateDepKind::Explicit
                 };
 
-                let cnum = self.resolve_crate(name, item.span, dep_kind);
+                self.used_extern_options.insert(name);
```

Comment:
Should the `used_extern_options` extension be included only in the error path? Otherwise we'd never complain about an `extern foo;` that exists, but is then never used, right?

Reply:
The behaviour of this line is equivalent to what it was before, when it was still calling `resolve_crate`
```rust
fn resolve_crate<'b>(
    &'b mut self,
    name: Symbol,
    span: Span,
    dep_kind: CrateDepKind,
) -> CrateNum {
    self.used_extern_options.insert(name);
    self.maybe_resolve_crate(name, dep_kind, None).unwrap_or_else(|err| {
        let missing_core =
            self.maybe_resolve_crate(sym::core, CrateDepKind::Explicit, None).is_err();
        err.report_and_throw(&self.sess, span, missing_core)
    })
}
```
So as of now, the crate will never complain. Do I understand right, that you are saying, it should only be marked as "used", in the error case, to surpress later warnings, but it should not be added in the happy case?

Reply:
You got what I meant right. I now wonder why it is always silenced... That *might* be fine 🤷‍♀️

Leaving it to petrochenkov to decide whether to change that 🙂

Reply:
Oddly enough, I tried to come up with an example that would show the difference in error output, but it doesn't seem to make a difference. The only test that would seem to break with the change would be [`[ui] ui\unused-crate-deps\use_extern_crate_2015.rs`](https://github.com/rust-lang/rust/blob/master/src/test/ui/unused-crate-deps/use_extern_crate_2015.rs).

Reply:
This is about command line `--extern` options, not about unused `extern crate` items.
So when we resolve `extern crate foo`, then `--extern foo` gets used regardless of the resolution results.